lib/internal/worker.js

```diff
@@ -0,0 +1,275 @@
+// in some other thread.
+class MessageDuplex extends EventEmitter {
+  constructor(recvHandle, send) {
+    super();
+    assert.strictEqual(typeof recvHandle, 'object');
+    assert.strictEqual(typeof send, 'function');
+    this[kRecvHandle] = recvHandle;
+    this[kRecvHandle].onmessage = (flag, buf) => this[kOnMessage](flag, buf);
+    this[kSend] = send;
+    debug(`[${process.threadId}] Created MessageDuplex wrapper object`);
+  }
+
+  // Lifetime management of workers is achieved through ref()/unref()ing
+  // their corresponding message channels.
+  ref() {
+    debug(`[${process.threadId}] MessageDuplex.ref()`);
+    this[kRecvHandle].ref();
+  }
+
+  unref() {
+    debug(`[${process.threadId}] MessageDuplex.unref()`);
+    this[kRecvHandle].unref();
+  }
+
+  // Called when a message is emitted by C++.
+  [kOnMessage](flag, buf) {
+    if (flag !== OOB_NONE && flag < OOB_CUSTOM_OFFSET) {
+      // This was not handled in C++, but it is also not a custom message in the
+      // sense that it was generated in JS, so some special handling is still
+      // required for deserialization.
+      // (This is primarily for error situations)
+      return this.emit('message', flag, buf);
+    }
+
+    const deserializer = new Deserializer(buf);
+    deserializer.readHeader();
+    const payload = deserializer.readValue();
+    debug(`[${process.threadId}] received message`, flag, payload);
+    // Emit the flag and deserialized object to userland.
+    this.emit('message', flag, payload);
+  }
+
+  write(flag, payload) {
+    debug(`[${process.threadId}] sending message`, flag, payload);
+    const serializer = new Serializer();
+    serializer.writeHeader();
+    serializer.writeValue(payload);
+    this[kSend](flag, serializer.releaseBuffer());
+  }
+}
+
+class Worker extends EventEmitter {
+  constructor(filename, options = {}) {
+    super();
+    debug(`[${process.threadId}] create new worker`, filename, options);
+    if (typeof filename !== 'string') {
+      throw new errors.TypeError('ERR_INVALID_ARG_TYPE', 'filename',
+                                 'string', filename);
+    }
+
+    const resourceLimits = {
+      maxSemiSpaceSize: options.maxSemiSpaceSize | 0,
+      maxOldSpaceSize: options.maxOldSpaceSize | 0
+    };
+
+    // Set up the C++ handle for the worker, as well as some internal wiring.
+    this[kHandle] = new WorkerImpl(resourceLimits);
+    this[kHandle].onexit = (code) => this[kOnExit](code);
+    this[kChannel] =
+        new MessageDuplex(this[kHandle].getChildChannel(),
+                          (flag, buf) => this[kHandle].send(flag, buf));
+    this[kChannel].on('message', (flag, payload) => {
+      this[kOnMessage](flag, payload);
+    });
+    this[kStartupScriptFilename] = filename;
+    debug(`[${process.threadId}] created Worker with ID ${this.threadId}`);
+
+    // Actually start the new thread now that everything is in place.
+    this[kHandle].startThread();
```

Comment:
I'm not sure if we shouldn't provide `.start()` as an external method so creating and starting the work are different, but I don't feel strongly about it.

Reply:
Yeah, it’s an arbitrary decision in my eyes, too. Right now, all communication with the Worker is asynchronous anyway, so I couldn’t find any advantage in adding a separate method, but if there is some, fine by me. :)